compiler/rustc_lint/src/mutable_transmutes.rs

```diff
@@ -0,0 +1,718 @@
+//! This module check that we are not transmuting `&T` to `&mut T` or `&UnsafeCell<T>`.
+//!
+//! The complexity comes from the fact that we want to lint against this cast even when the `&T`, the `&mut T`
+//! or the `&UnsafeCell<T>` (either the `&` or the `UnsafeCell` part of it) are hidden in fields.
+//!
+//! The general idea is to first isolate potential candidates, then check if they are really problematic.
+//!
+//! That is, we first collect all instances of `&mut` references in the target type, then we check if any
+//! of them overlap with a `&` reference in the source type.
+//!
+//! We do a similar thing to `UnsafeCell`. Here it is a little more complicated, since we operate two layers deep:
+//! first, we collect all `&` references in the target type. Then, we check if any of them overlaps with a `&` reference
+//! in the source type (not a `&mut` reference, since it is perfectly fine to transmute `&mut T` to `&UnsafeCell<T>`).
+//! If we found such overlap, we collect all instances of `UnsafeCell` under the reference in the target type.
+//! If we found any, we try to find if it overlaps with things that are not `UnsafeCell` under the reference
+//! in the source type.
+
+use std::fmt::Write;
+
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::{BorrowKind, Expr, ExprKind, UnOp};
+use rustc_index::IndexSlice;
+use rustc_middle::bug;
+use rustc_middle::ty::layout::LayoutOf;
+use rustc_middle::ty::{self, AdtDef, GenericArgsRef, List, Mutability, Ty};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::symbol::sym;
+use rustc_target::abi::{FieldIdx, FieldsShape, HasDataLayout, Layout, Size};
+use tracing::{debug, debug_span};
+
+use crate::lints::{BuiltinMutablesTransmutes, TransmuteBreadcrumbs, UnsafeCellTransmutes};
+use crate::{LateContext, LateLintPass, LintContext};
+
+fn struct_fields<'a, 'tcx>(
+    cx: &'a LateContext<'tcx>,
+    layout: Layout<'tcx>,
+    adt_def: AdtDef<'tcx>,
+    substs: GenericArgsRef<'tcx>,
+) -> impl Iterator<Item = (FieldIdx, Size, Ty<'tcx>)> + 'a {
+    let field_tys =
+        adt_def.non_enum_variant().fields.iter().map(|field| {
+            cx.tcx.normalize_erasing_regions(cx.typing_env(), field.ty(cx.tcx, substs))
+        });
+    // I thought this could panic, but apparently SIMD vectors are structs but also have primitive layout. So...
```

Comment:
It's unclear what "this" refers to, which makes the comment hard for me to understand.

Reply:
Fixed. It refers to the following `match`.